### DISCUSSION TITLE
In-memory DB compatible with MikroORM and NestJS for integration testing (API endpoints)

### DISCUSSION THREAD

**Discussion by Hiroki111** [mikro-orm/mikro-orm#5406] [General]
**Is your feature request related to a problem? Please describe.**
I want to be able to test whether API endpoints (built with NestJS) return an expected response body based on what the DB currently stores.

https://mikro-orm.io/docs/guide/project-setup#testing-the-endpoint
I followed this example to see if I can use an in-memory DB for testing. By following it, I thought`MikroORM.init` would be called with
```
{
  entities: [/* entities here*/],
  dbName: ':memory:',
  debug: false,
  driver: PostgreSqlDriver, // I use Postgres
};
```
(I added entities and driver fields, because omitting them seems to cause the following errors:

> No entities found, please use `entities` option

> No driver specified, please fill in the `driver` option or use `defineConfig` helper (to define your ORM config) or `MikroORM` class (to call the `init` method) exported from the driver package (e.g. `import { defineConfig } from '@mikro-orm/mysql'; export defineConfig({ ... })`)

However, using the above config data resulted in "connect ECONNREFUSED 127.0.0.1:5432". So, I feel that MikroORM can use an in-memory DB only with Fastify, which is used in the example above, not with NestJS.

**Describe the solution you'd like**

Ideally, I want to be able to test API endpoints in the following way:

```
import { MikroOrmModuleOptions } from '@mikro-orm/nestjs/typings';
import * as request from 'supertest';

const testMikroOrmModuleOptions: MikroOrmModuleOptions = {
  dbName: ':memory:',
  debug: false,
};

describe('My controller', () => {
  let app: INestApplication;
  let entityManager: EntityManager;

  beforeAll(async () => {
    const testingModule = await Test.createTestingModule({
      imports: [AppModule],
    })
    .overrideProvider(MikroORMConfig) // Actuall ORM config, created by MikroOrmModule.forRoot(MikroOrmModuleOptions)
    .useValue(MikroOrmModule.forRoot(testMikroOrmModuleOptions)) // This is the config used for testing
    .compile();
    app = testingModule.createNestApplication();
    entityManager = testingModule.get<EntityManager>(EntityManager);
    await app.init();
  });

  afterEach(() => {
    entityManager.clear();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /foo', () => {
    it('should return foo', async () => {
      const mockFoo = {/** data of mockFoo */};
      // populate the in-memory DB with mockFoo via entityManager

      return request(app.getHttpServer())
        .get(`/foo`)
        .expect(200)
        .expect('Content-Type', /json/)
        .then((response) => {
          expect(response.body).toBe([mockFoo]);
        });
    });
  });
});

```

**Describe alternatives you've considered**
Perhaps I could create a normal DB instance dedicated to integration tests, rather than an in-memory DB. However, tests would be slow to execute and it'll take additional DB space, neither of which is ideal

**Comment by B4nan** [comment]
This is out of scope of this project, reimplementing SQL in memory would be quite a huge effort. Also, I don't see how it would be any better than just using sqlite directly. If you don't want to run your tests against a real database, you can use sqlite driver only for tests. It will impose some limitations on what you can use on schema level (could be worked around via metadata discovery hooks), but using some custom made in memory adapter would be much worse in this regard.

**Reply by Hiroki111** [reply]
> I don't see how it would be any better than just using sqlite directly. If you don't want to run your tests against a real database, you can use sqlite driver only for tests.

I get this point. The issue that I'm facing is that my NestJS application has DB migration files written for Postgres. These files use SQL syntax that is incompatible with Sqlite

**Reply by B4nan** [reply]
You don't have to use migrations for tests, it's more about depending on the dialect-specific features, e.g. column defaults or SQL expressions on various places. As long as you know the schema is in sync, you can just do `orm.schema.createSchema()` and you should be good to go.

Anyway, as I said, what you are asking for would end up as much more limiting than using sqlite. If sqlite is not an option for you, better to just use the real database.

There is also https://github.com/oguimbal/pg-mem which even provides an adapter for MikroORM, but its also very limited, it wouldn't work for the ORM test suite at all.

**Reply by Hiroki111** [reply]
@B4nan 
I don't see if `orm.schema.createSchema()` creates tables.

If I run the following...

```
import { SqliteDriver } from '@mikro-orm/sqlite'
import { MikroOrmModule } from '@mikro-orm/nestjs';;
import { MikroOrmModuleOptions } from '@mikro-orm/nestjs/typings';
import { EntityManager } from '@mikro-orm/sqlite';
import { MikroORM } from '@mikro-orm/core';
import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
// other dependencies here

const mikroOrmModuleOptionsForTesting: MikroOrmModuleOptions = {
  entities: [/** entities */],
  dbName: ':memory:',
  debug: false,
  driver: SqliteDriver,
  allowGlobalContext: true, // I know I shouldn't use it. I just wanted to temporarily use it for experimentation purposes
};

describe('My tests', () => {
  let app: INestApplication;
  let orm: MikroORM;
  let entityManager: EntityManager;

  beforeAll(async () => {
    const testingModule = await Test.createTestingModule({
      imports: [
        MikroOrmModule.forRoot(mikroOrmModuleOptionsForTesting),
        MyModule, // The modules that I need to test
      ],
    })
    .compile();
    app = testingModule.createNestApplication();
    orm = await MikroORM.init(mikroOrmModuleOptionsForTesting);
    entityManager = orm.em as EntityManager;
    await orm.schema.createSchema();
    await app.init();
  });
  
  // test cases here
```

... and examine "entityManager" by console.log
```
import { EntityManager } from '@mikro-orm/postgresql';
import { Injectable } from '@nestjs/common';

@Injectable()
export class MyService {
  constructor(
    public readonly entityManager: EntityManager,
    private readonly MyService: MyService,
  ) {
    console.log('entityManager', this.entityManager.getMetadata());
    // If I put "await this.entityManager.findAll(MyEntity))" here, it throws an error
  }
```

I can see that `this.entityManager` contains the entity that I use. However, trying to get data from the DB table by `this.entityManager.findAll(MyEntity)` results in `SQLITE_ERROR: no such table: my_entity` (where "my_entity" is the actual table name)

Can this be possibly be caused by Sqlite?

**Reply by B4nan** [reply]
Your nest app will create its own `MikroORM` instance, it won't use the one you create explicitly, there is no link - so the `createSchema` call happens on a different in memory database than the one nest uses.

(I'd guess)

**Reply by B4nan** [reply]
Try getting it from the nest app, something like `app.get(MikroORM)` instead.

**Reply by Hiroki111** [reply]
I finally found a way to make things work, thank you!

**Reply by andreciobezerra** [reply]
Please, @Hiroki111, may you share the solution ?
